Sources/Common/Background Queue/QueueRunRequest.swift

```diff
@@ -135,11 +135,17 @@ public class CioQueueRunRequest: QueueRunRequest {
                         self.logger.info("queue is quitting early because all HTTP requests are paused.")
 
                         doneRunning()
-                    } else if case .badRequest = error {
-                        self.logger.debug("queue task \(nextTaskStorageId) failed with 400")
+                    } else if case .badRequest400 = error {
+                        self.logger.debug("queue deleting task \(nextTaskStorageId) because it failed with 400 and will always fail")
 
-                        self.logger.debug("queue deleting task \(nextTaskStorageId) because it will always fail")
-                        _ = self.storage.delete(storageId: nextTaskToRunInventoryItem.taskPersistedId)
+                        self.logger.error("Received HTTP 400 response while trying to <queue task type>. 400 responses never succeed and therefore, the SDK is deleting this SDK request and not retry. Error message from API: <error message string>, request data sent: <queue task data JSON>")
+
+                        // check if its a groupStart task, if so delete the group otherwise just delete the task
+                        if let groupStart = nextTaskToRunInventoryItem.groupStart {
+                            _ = self.storage.deleteGroup(groupStartTask: groupStart)
+                        } else {
+                            _ = self.storage.delete(storageId: nextTaskToRunInventoryItem.taskPersistedId)
+                        }
```

Comment:
I think this logic would benefit from belonging somewhere else. `QueueRunRequest` is large and adding more logic in it can make even it harder to test. I think moving this logic into `QueueStorage` itself would help that. 

Some ideas that we could do. 

### Modify `delete()`

We could modify `QueueStorage.delete` to take a new optional parameter. 

```
func delete(storageId: String, deleteGroupIfStartOfGroup: Bool = false) -> Bool {
  // delete the task. 

  if deleteGroupIfStartOfGroup && task.groupStart != nil {
    // delete all tasks in group
  }
} 
```

### Add a function to `QueueStorage` to do this logic 

Add a function `QueueStorage.deleteGroupIfStartOfGroup(storageId: )` that does the logic here in this file. 

---

This makes the test for `QueueRunRequest` easier since we just need to confirm that when a 400 happens, we call a specific function on `QueueStorage`, but we don't need to worry about the logic of that function in `QueueStorage`. 

I don't have a preference for the solution. If you have better ideas for naming I am open.

Reply:
Conclusion after sync: 

// delete task
// conditional check if group, then delete group